lib/darwin/sys/proctable.rb

```diff
@@ -273,6 +274,18 @@ def self.get_cmd_args_and_env(pid, struct)
       # Since we may not have access to the process information due
       # to improper privileges, just bail if we see a failure here.
 
+      # First use KERN_PROCARGS2 to discover the argc value of the running process.
+      mib.write_array_of_int([CTL_KERN, KERN_PROCARGS2, pid])
+      return if sysctl(mib, 3, nil, len, nil, 0) < 0
+
+      buf = FFI::MemoryPointer.new(:char, len.read_ulong)
+      return if sysctl(mib, 3, buf, len, nil, 0) < 0
+
+      # The argc value is located in the first byte of buf
+      argc = buf.read_bytes(1).ord
+      buf.free
+
```

Comment:
Since this is a one time buffer, I think the block form would be appropriate here.

    FFI::MemoryPointer.new(:char, len.read_ulong) do |buf|
        ...
    end